File: C++/Subarray_Array_Medium_MaximumProductSubarray_152.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) 
    {
        int maxCount = nums[0];
        int count_p = nums[0];
        int count_n = nums[0];

        for(int i = 1; i < nums.size(); i++)
        {
            int curr = nums[i];
            int temp = count_p;

            count_p = max(curr, max(temp*curr, count_n*curr));
            count_n = min(curr, min(temp*curr, count_n*curr));

            maxCount = max(count_p, maxCount);
        }

        return maxCount;
    }
};

File: Python/Subarray_Array_Medium_MaximumProductSubarray_152.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        
        maxCount = nums[0]
        count_p = nums[0]
        count_n = nums[0]

        for i in range(1, len(nums)):

            curr = nums[i]

            temp = count_p

            count_p = max(curr, temp * curr, count_n * curr)
            count_n = min(curr, temp * curr, count_n * curr)

            maxCount = max(count_p, maxCount)

        return maxCount
        